File: src/query/context_builder.py
"""Montagem de contexto para o LLM e guardrail de veracidade.

Decisao de design:
    O contexto para o LLM e uma unica string com blocos ``[Fonte: titulo (data) - url]``
    concatenados por ``---``. Nao usamos listas/objetos porque a string sera
    injetada literalmente no prompt.

Coexistencia com Task 6.1:
    Este modulo depende apenas de ``ScoredChunk`` (Task 2.2) e de
    ``MIN_RELEVANCE_SCORE`` (Task 6.1). NAO importa ``QueryEngine`` para nao
    travar o modulo caso Task 6.1 ainda esteja em andamento.
"""
from __future__ import annotations

from src.db.vector_store import ScoredChunk
from src.query.constants import MIN_RELEVANCE_SCORE


NO_EVIDENCE_MESSAGE: str = "Nao encontrei base para responder"

_BLOCK_SEP: str = "\n---\n"


def _format_published_at(published_at) -> str:
    """Formata published_at como ' publicado em YYYY-MM-DD' ou '' se None."""
    if published_at is None:
        return ""
    return f" publicado em {published_at.strftime('%Y-%m-%d')}"


def build_context(ranked: list[ScoredChunk]) -> str:
    """Monta contexto para o LLM a partir de chunks ranqueados.

    Cada bloco tem o formato::

        [Fonte: {doc_title} - publicado em {YYYY-MM-DD} - {source_url}]
        {text}

    Quando ``published_at`` e None, o segmento de data e omitido (evita
    aparecer como ' - - ' no bloco). Os blocos sao juntados por ``---``.

    A data de publicacao ajuda o LLM a preferir a versao mais recente quando
    ha conflitos entre versoes de uma mesma NT.

    Ordenacao: respeita a ordem recebida (chamador ja ranqueou por score).
    """
    if not ranked:
        return ""
    blocks: list[str] = []
    for c in ranked:
        date_segment = _format_published_at(c.published_at)
        if date_segment:
            header = f"[Fonte: {c.doc_title} -{date_segment} - {c.source_url}]"
        else:
            header = f"[Fonte: {c.doc_title} - {c.source_url}]"
        blocks.append(f"{header}\n{c.text}\n")
    return _BLOCK_SEP.join(blocks)


def has_sufficient_evidence(
    ranked: list[ScoredChunk],
    min_score: float = MIN_RELEVANCE_SCORE,
) -> bool:
    """Guardrail de veracidade: retorna ``True`` apenas se ha pelo menos 1 chunk
    com ``score >= min_score``.

    Convencao: olhamos apenas o PRIMEIRO chunk (mais relevante no ranking).
    Se ele nao atinge o minimo, qualquer outro posterior tera score menor, entao
    a checagem e correta sem percorrer a lista inteira.
    """
    return len(ranked) > 0 and ranked[0].score >= min_score


__all__ = [
    "NO_EVIDENCE_MESSAGE",
    "build_context",
    "has_sufficient_evidence",
]

File: packages/dfe-agent/src/query/contextBuilder.ts
/**
 * contextBuilder.ts — monta o contexto final (answer + sources).
 *
 * @see PLAN_SPRINT14.md Task E.5
 *
 * Constante NO_EVIDENCE_MESSAGE canonica em `src/query/index.ts` (mesma
 * string em src/query/context_builder.py do Py). NUNCA duplique o literal
 * em outro arquivo — gate dfe-rules.md #4 + convencoes-gerais.md.
 */

import { NO_EVIDENCE_MESSAGE, MIN_RELEVANCE_SCORE } from "./constants.js";
import type { Source } from "./index.js";

export interface HydratedChunk {
  chunk_id: number;
  doc_id: number;
  text: string;
  url: string;
  title: string;
  score: number;
  published_at?: string | null;
}

export interface ContextResult {
  answer: string;
  sources: Source[];
}

/**
 * Heuristica de evidencia suficiente (gate dfe-rules.md #4 + paridade Py).
 *
 * Convencao Py (`src/query/context_builder.py:60-71`): olhamos apenas o
 * PRIMEIRO chunk (mais relevante no ranking). Se ele nao atinge o minimo,
 * qualquer outro posterior tera score menor.
 *
 * Antes da Sprint 14.1, Node usava `Math.max(...)` que retornava True em
 * cenarios com 1 chunk bom + N chunks ruins, divergindo do Py. Corrigido
 * para paridade.
 */
export function hasSufficientEvidence(chunks: HydratedChunk[]): boolean {
  if (chunks.length === 0) return false;
  return chunks[0].score >= MIN_RELEVANCE_SCORE;
}

/**
 * Constroi o contexto final: dedup por URL, ordena por score, monta answer.
 */
export function buildContext(chunks: HydratedChunk[]): ContextResult {
  if (!hasSufficientEvidence(chunks)) {
    return {
      answer: NO_EVIDENCE_MESSAGE,
      sources: [],
    };
  }

  // Dedup por URL (preferir chunk com maior score)
  const byUrl = new Map<string, HydratedChunk>();
  for (const c of chunks) {
    const existing = byUrl.get(c.url);
    if (!existing || c.score > existing.score) {
      byUrl.set(c.url, c);
    }
  }

  const sorted = Array.from(byUrl.values()).sort((a, b) => b.score - a.score);

  const sources: Source[] = sorted.map((c) => ({
    url: c.url,
    title: c.title,
    score: c.score,
  }));

  const answer = sorted.map((c) => c.text).join("\n\n---\n\n");

  return { answer, sources };
}